libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,347 @@
+}
+
+FunctionCallGraphBuilder::ContractCallGraph FunctionCallGraphBuilder::buildCreationGraph(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder(_contract);
+	solAssert(builder.m_currentNode == Node(SpecialNode::Entry), "");
+
+	// Create graph for constructor, state vars, etc
+	for (ContractDefinition const* base: _contract.annotation().linearizedBaseContracts | views::reverse)
+	{
+		// The constructor and functions called in state variable initial assignments should have
+		// an edge from Entry
+		builder.m_currentNode = SpecialNode::Entry;
+		for (auto const* stateVar: base->stateVariables())
+			stateVar->accept(builder);
+
+		if (base->constructor())
+		{
+			builder.functionReferenced(*base->constructor());
+
+			// Constructors and functions called in state variable initializers have an edge either from
+			// the previous class in linearized order or from Entry if there's no class before.
+			builder.m_currentNode = base->constructor();
+		}
+
+		// Functions called from the inheritance specifier should have an edge from the constructor
+		// for consistency with functions called from constructor modifiers.
+		for (auto const& inheritanceSpecifier: base->baseContracts())
+			inheritanceSpecifier->accept(builder);
+	}
+
+	builder.m_currentNode = SpecialNode::Entry;
+	builder.processQueue();
+
+	return move(builder.m_graph);
+}
+
+FunctionCallGraphBuilder::ContractCallGraph FunctionCallGraphBuilder::buildDeployedGraph(
+	ContractDefinition const& _contract,
+	FunctionCallGraphBuilder::ContractCallGraph const& _creationGraph
+)
+{
+	solAssert(&_creationGraph.contract == &_contract, "");
+
+	FunctionCallGraphBuilder builder(_contract);
+	solAssert(builder.m_currentNode == Node(SpecialNode::Entry), "");
+
+	auto getSecondElement = [](auto const& _tuple){ return get<1>(_tuple); };
+
+	// Create graph for all publicly reachable functions
+	for (FunctionTypePointer functionType: _contract.interfaceFunctionList() | views::transform(getSecondElement))
+	{
+		auto const* function = dynamic_cast<FunctionDefinition const*>(&functionType->declaration());
+		auto const* variable = dynamic_cast<VariableDeclaration const*>(&functionType->declaration());
+
+		if (function)
+			builder.functionReferenced(*function);
+		else
+			// If it's not a function, it must be a getter of a public variable; we ignore those
+			solAssert(variable, "");
+	}
+
+	if (_contract.fallbackFunction())
+		builder.functionReferenced(*_contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		builder.functionReferenced(*_contract.receiveFunction());
+
+	// All functions present in internal dispatch at creation time could potentially be pointers
+	// assigned to state variables and as such may be reachable after deployment as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	set<Node, CompareByID> emptySet;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, emptySet))
```

Comment:
Why not?
```suggestion
	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, set<Node, CompareByID>{}))
```
This has the advantage that the variable cannot be referenced later on.

Reply:
It does seem to work for me (even though it returns a temporary) and would be my preferred solution but [@bshastry reported that asan complains about `stack-use-after-scope`](https://github.com/ethereum/solidity/pull/10973#discussion_r578339963).

Reply:
That is weird - isn't its lifetime extended to the end of the range-based for loop?

Reply:
Might be a false-positive due to the fact that I am passing `dispatchTarget` into a function which does actually stored it. It's safe only because the set is empty and asan might not be accounting for that.

Should I just use the temporary and ignore the potential warning? The old version wasn't failing in CI so it would work.

There are also two other uses of `valueOrDefault()` which @bshastry did not point out so I guess there were no warnings for them. I changed them along with this one but if we think it's safe, I could change back at least those two places.

Reply:
I have just pushed two fixups that restore the temporaries.

Oh, and it was actually one other place that uses `valueOrDefault()`, not two. Looks like removing the BFS eliminated the one in `IRGenerator.cpp`.